cmd/juju/application/diffbundle.go

```diff
@@ -265,15 +270,40 @@ func (c *bundleDiffCommand) bundleDataSource(ctx *cmd.Context) (charm.BundleData
 	return store.NewResolvedBundle(bundle), nil
 }
 
-func (c *bundleDiffCommand) charmAdaptor() (BundleResolver, error) {
-	if c._charmStore != nil {
-		return c._charmStore, nil
-	}
-	controllerName, err := c.ControllerName()
-	if err != nil {
-		return nil, errors.Trace(err)
+func (c *diffBundleCommand) charmAdaptor(apiRoot base.APICallCloser, curl *charm.URL) (BundleResolver, error) {
+	var resolver BundleResolverFactory
+	switch curl.Schema {
+	case "cs":
+		resolver = charmStoreBundleResolver{
```

Comment:
Why not call to a c-tor and return something that implements BundleResolver? A `charmStoreBundleResolver` sounds like it should implement the interface given the way it is named. This way you can get rid of the factory interface below.